engine/flink/components/base/src/main/scala/pl/touk/nussknacker/engine/flink/util/transformer/join/FullOuterJoinTransformer.scala

```diff
@@ -0,0 +1,162 @@
+            agg.computeOutputType(aggregateByByBranchId(id).returnType)
+              .leftMap(x => {
+                val branchParamId = ParameterNaming.getNameForBranchParameter(AggregateByParam.parameter, id)
+                NonEmptyList(CustomNodeError(x, Some(branchParamId)), Nil)})
+              .map(id -> _)
+          }.toList.sequence.map(_.toMap)
+          validatedAggregatorReturnTypes.map(outputTypeByBranchId => {
+            val outputTypes = outputTypeByBranchId.toList
+              .map{case (k, v) => ContextTransformation.sanitizeBranchName(k) -> v}
+              .++(List(KeyFieldName -> Typed.typedClass[String]))
+            TypedObjectTypingResult(outputTypes, Typed.typedClass[java.util.Map[String, AnyRef]])
+          })
+
+        case _ => Validated.validNel(Unknown)
+      }
+
+      FinalResults.forValidation(mainCtx, validatedOutputType.swap.map(_.toList).getOrElse(List()))(
+        _.withVariable(OutputVar.customNode(outName), validatedOutputType.getOrElse(Unknown)))
+  }
+
+  override def implementation(params: Map[String, Any], dependencies: List[NodeDependencyValue], finalState: Option[State]): FlinkCustomJoinTransformation = {
+    val keyByBranchId: Map[String, LazyParameter[CharSequence]] = KeyParam.extractValue(params)
+    val aggregatorByBranchId: Map[String, Aggregator] = AggregatorParam.extractValue(params)
+    val aggregateByByBranchId: Map[String, LazyParameter[AnyRef]] = AggregateByParam.extractValue(params)
+    val window: Duration = WindowLengthParam.extractValue(params)
+
+    val aggregator: Aggregator = new MapAggregator(
+      aggregatorByBranchId.mapValues(new OptionAggregator(_).asInstanceOf[Aggregator]).asJava)
+
+    val baseElement: Map[String, AnyRef] = keyByBranchId.keySet.map(_ -> None).toMap
+
+    (inputs: Map[String, DataStream[Context]], context: FlinkCustomNodeContext) => {
+      val keyedStreams = inputs.map {
+        case (id, stream) =>
+          stream.flatMap(new StringKeyedValueMapper(context, keyByBranchId(id), aggregateByByBranchId(id)))
+            .map(_.map(_.mapValue(x => {
+              val sanitizedId = ContextTransformation.sanitizeBranchName(id)
+              (baseElement + (sanitizedId -> Some(x))).asJava.asInstanceOf[AnyRef]
+            })))
+      }
+
+      val types = aggregateByByBranchId.mapValues(_.returnType)
+      val optionTypes = types.mapValues(t => Typed.genericTypeClass(classOf[Option[_]], List(t)))
+      val inputType = TypedObjectTypingResult(optionTypes.toList, objType = Typed.typedClass[java.util.Map[_, _]])
+
+      val storedType = aggregator.computeStoredTypeUnsafe(inputType)
+      val storedTypeInfo = context.typeInformationDetection.forType(storedType)
+      val aggregatorFunction = prepareAggregatorFunction(aggregator, FiniteDuration(window.toMillis, TimeUnit.MILLISECONDS), inputType, storedTypeInfo, context.convertToEngineRuntimeContext)(NodeId(context.nodeId))
+
+      val stream = keyedStreams
+        .reduce(_.union(_))
```

Comment:
It will work as expected because sustaining state is independent of incoming messages, it's always sustained for set amount of time. Test "timeouts" whould behave differently if state was actualy sustained by new events.